internal/controllers/flowcontrol/synthesis.go

```diff
@@ -0,0 +1,81 @@
+package flowcontrol
+
+import (
+	"context"
+	"fmt"
+
+	apiv1 "github.com/Azure/eno/api/v1"
+	"github.com/Azure/eno/internal/manager"
+	"github.com/go-logr/logr"
+	"github.com/google/uuid"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type synthesisConcurrencyLimiter struct {
+	client client.Client
+	limit  int
+}
+
+func NewSynthesisConcurrencyLimiter(mgr ctrl.Manager, limit int) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("synthesisConcurrencyLimiter").
+		Watches(&apiv1.Composition{}, manager.SingleEventHandler()).
+		WithLogConstructor(manager.NewLogConstructor(mgr, "synthesisConcurrencyLimiter")).
+		Complete(&synthesisConcurrencyLimiter{
+			client: mgr.GetClient(),
+			limit:  limit,
+		})
+}
+
+func (c *synthesisConcurrencyLimiter) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := logr.FromContextOrDiscard(ctx)
+
+	list := &apiv1.CompositionList{}
+	err := c.client.List(ctx, list)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	var active int
+	var pending int
+	var next *apiv1.Composition
+	for _, comp := range list.Items {
+		comp := comp
+		current := comp.Status.CurrentSynthesis
+		if current != nil && current.Synthesized != nil {
+			continue // already synthesized
+		}
+		if current == nil || current.UUID == "" {
+			pending++
+			next = &comp
```

Comment:
Should we sort (or shuffle) to have better guarantees around choosing the next comp to be dispatched?

Reply:
There's a risk of the loop "outpacing" the informer, since it will be re-entered by any composition change and not just changes caused by the controller. Currently it will fail closed by retrying the same composition and getting a conflict error, shuffling would potentially result in a few extras getting dispatched before the informer catches up.

I do think this is worth addressing, but I'll do so in a separate PR to keep the scope of this one somewhat reasonable